From f763f1884bc9e364699da01e928041498209accf Mon Sep 17 00:00:00 2001
From: jane400 <pmos@j4ne.de>
Date: Fri, 21 Jun 2024 22:27:39 +0200
Subject: [PATCH] pmos/systemd: introduce $systemd_system_services and
 $systemd_user_services

Syntax is an whitespace seperated list of units for both.

This is done by appending the commands to the .post-install,
.pre-deinstall hooks. Those hooks will get created on the fly during
package creating if the APKBUILD didn't provide it.

Settings this will run `systemctl preset`
---
 abuild.in | 110 ++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 110 insertions(+)

diff --git a/abuild.in b/abuild.in
index 576ef2d..990abc0 100644
--- a/abuild.in
+++ b/abuild.in
@@ -1116,6 +1116,79 @@ check_depver() {
 	return 0
 }
 
+parse_systemd_services() {
+	local system_preset=""
+	local user_preset=""
+
+	for i in $systemd_system_services; do
+		local n=${i%:*}
+		local suff=${i##*:}
+		if [ ! -e "$dir/usr/lib/systemd/system/$n" ]; then
+			continue
+		fi
+		system_preset="$system_preset $n"
+	done
+
+	for i in $systemd_user_services; do
+		local n=${i%:*}
+		local suff=${i##*:}
+		if [ ! -e "$dir/usr/lib/systemd/user/$n" ]; then
+			continue;
+		fi
+		user_preset="$user_preset $n"
+	done
+
+	if [ -n "$system_preset$user_preset" ]; then
+		local systemd_functions
+		systemd_functions="
+SYSTEMD_BUS_TIMEOUT=5s
+have_systemctl() {
+	command -v systemctl > /dev/null
+}
+is_systemd_running() {
+	[ -d "/run/systemd/running" ]
+}
+"
+		append_post_install="$append_post_install$systemd_functions"
+		append_pre_deinstall="$append_pre_deinstall$systemd_functions"
+	fi
+
+	if [ -n "$system_preset" ]; then
+		append_post_install="$append_post_install
+have_systemctl && systemctl --no-reload preset $system_preset || true"
+		append_pre_deinstall="$append_pre_deinstall
+have_systemctl && if is_systemd_running; then
+	systemctl --no-reload disable --now --no-warn $system_preset
+else
+	systemctl --no-reload disable --no-warn $system_preset
+fi || true"
+	fi
+
+	if [ -n "$user_preset" ]; then
+		append_post_install="$append_post_install
+have_systemctl && systemctl --no-reload preset --global $user_preset || true"
+		append_pre_deinstall="$append_pre_deinstall
+if have_systemctl; then
+	systemctl --global disable --no-warn "$@"
+
+	if is_systemd_running; then
+		users=$(systemctl list-units 'user@*' --legend=no | sed -n -r 's/.*user@([0-9]+).service.*/\1/p')
+		for user in $users; do
+			systemctl --user -M "$user@" disable --now --no-warn "$@" &
+	    done
+	    wait
+	fi
+
+fi || true
+"
+	fi
+
+	# add the systemctl dep
+	if [ -n "$user_preset$system_preset" ]; then
+		depends="$depends cmd:systemctl /bin/sh"
+	fi
+}
+
 prepare_metafiles() {
 	local name=${subpkgname:-$pkgname}
 	[ -z "${name##* *}" ] && die "package name contains spaces"
@@ -1135,6 +1208,15 @@ prepare_metafiles() {
 		sync;;
 	esac
 
+	local append_post_install
+	local append_pre_deinstall
+
+	# the systemd package doesn't split the service files into -service subpackages
+	# as they're heavily intertwined with the functioning of systemd.
+	if [ -z "${subpkgname##*-service}"] || [ "$pkgname" = "systemd" ]; then
+		parse_systemd_services
+	fi
+
 	local size=$(du -sk | awk '{print $1 * 1024}')
 	# If package contains only empty files (or only install scripts), the size
 	# might be 0. But due to apk-tools 2 considering packages with size = 0
@@ -1239,6 +1321,34 @@ prepare_metafiles() {
 		chmod +x "$controldir/$script"
 		metafiles="$metafiles $script"
 	done
+
+	if [ -n "$append_post_install" ]; then
+		script=".post-install"
+		if [ ! -f "$controldir/$script" ]; then
+			msg "Adding auto-generated $script"
+			echo "#!/bin/sh" > "$controldir/$script"
+			chmod +x "$controldir/$script"
+			metafiles="$metafiles $script"
+		fi
+		append_post_install="
+# The following part was autogenerated by abuild
+$append_post_install"
+		echo "$append_post_install" >> "$controldir/$script"
+	fi
+	if [ -n "$append_pre_deinstall"	 ]; then
+		script=".pre-deinstall"
+		if [ ! -f "$controldir/$script" ]; then
+			msg "Adding auto-generated $script"
+			echo "#!/bin/sh" > "$controldir/$script"
+			chmod +x "$controldir/$script"
+			metafiles="$metafiles $script"
+		fi
+		append_pre_deinstall="
+# The following part was autogenerated by abuild
+$append_pre_deinstall"
+		echo "$append_pre_deinstall" >> "$controldir/$script"
+	fi
+
 	echo $metafiles | tr ' ' '\n' > "$controldir"/.metafiles
 }
 
-- 
2.45.2

